### PR TITLE
fix: Validate if the GitHub installation is valid

### PR DESCRIPTION
**Pull request by novakzaballa**
Thanks for submitting a PR! Please check the boxes below:

- [x] I have run [`pre-commit`](https://docs.flagsmith.com/platform/contributing#pre-commit) to check linting
- [x] I have added information to `docs/` if required so people know about the feature!
- [x] I have filled in the "Changes" section below?
- [x] I have filled in the "How did you test this code" section below?
- [x] I have used a [Conventional Commit](https://www.conventionalcommits.org/en/v1.0.0/) title for this Pull Request

## Changes

- Add an infoMessage if the GitHub installation is invalid.
- Add the option to update the installation id or delete the integration, if it is invalid.
- Correct issue when try link a GitHub PR.
- Add extra logic in the GitHubSetupPage

## How did you test this code?

- Create an integration with GH in the Flagsmith App.
- Uninstall the app in GitHub
- Go to integration option in the side menu -> GitHub Integration -> Manage integration

### CODE REVIEW

File: frontend/web/components/App.js [hunk 1]
```diff
@@ -53,6 +56,7 @@ const App = class extends Component {
     getBuildVersion()
     this.listenTo(ProjectStore, 'change', () => this.forceUpdate())
     this.listenTo(AccountStore, 'change', this.getOrganisationUsage)
+    this.getGithubIntegration()
```
Comment (kyle-ssg) [inline frontend/web/components/App.js:59]: Can we move GH stuff in app to a component. App needs to have less noise

File: frontend/web/components/GitHubRepositoriesSelect.tsx [hunk 2]
```diff
@@ -73,6 +77,30 @@ const GitHubRepositoriesSelect: FC<GitHubRepositoriesSelectType> = ({
...
+          <Button
+            className='mr-3'
+            id='open-github-win-installations-btn'
+            data-test='open-github-win-installations-btn'
+            onClick={manageIntegration}
+            size='small'
+          >
+            Manage available GitHub Repositories
+          </Button>
+          <DeleteGithubIntegracion
```
Comment (kyle-ssg) [inline frontend/web/components/GitHubRepositoriesSelect.tsx:95]: DeleteGithubIntegration

File: frontend/web/components/GitHubRepositoriesSelect.tsx [hunk 3]
```diff
@@ -73,6 +77,30 @@ const GitHubRepositoriesSelect: FC<GitHubRepositoriesSelectType> = ({
...
+            githubId={githubId}
+            organisationId={organisationId}
+            onConfirm={() => {
```
Comment (kyle-ssg) [inline frontend/web/components/GitHubRepositoriesSelect.tsx:98]: onComplete I assume

File: frontend/web/components/modals/InvalidGithubIntegration.tsx [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+      <div className='modal-body'>
+        <ErrorMessage error={errorMessage} />
+        <InputGroup
+          title='New Installation ID'
+          inputProps={{
+            className: 'full-width',
+            name: 'installationId',
+          }}
+          value={installationId}
+          onChange={(event: React.ChangeEvent<HTMLInputElement>) => {
```
Comment (kyle-ssg) [inline frontend/web/components/modals/InvalidGithubIntegration.tsx:48]: I really don't understand what this is doing, why would we be typing an installation id ?

File: frontend/web/components/modals/CreateEditIntegrationModal.js [hunk 5]
```diff
@@ -233,29 +231,8 @@ const CreateEditIntegration = class extends Component {
...
-                  >
-                    Manage available GitHub Repositories
-                  </Button>
-                  <DeleteGithubIntegracion
-                    githubId={this.props.githubMeta.githubId}
-                    organisationId={AccountStore.getOrganisation().id}
-                    onConfirm={() => {
-                      closeModal()
-                    }}
+                    manageIntegration={this.openGitHubWinInstallations}
```
Comment (kyle-ssg) [inline frontend/web/components/modals/CreateEditIntegrationModal.js:234]: openGitHubWithInstallations

File: frontend/web/components/InfoMessage.js [hunk 6]
```diff
@@ -8,7 +8,11 @@ export default class InfoMessage extends PureComponent {
   static displayName = 'InfoMessage'
 
   handleOpenNewWindow = () => {
-    window.open(this.props.url, '_blank')
+    if (this.props.goToIntegrations) {
```
Comment (kyle-ssg) [inline frontend/web/components/InfoMessage.js:11]: Let's come up with a better solution than adding goToIntegrations as a prop to a generic component like InfoMessage, it doesn't make sense.

File: frontend/web/components/App.js [hunk 7]
```diff
@@ -474,6 +496,28 @@ const App = class extends Component {
                               id={AccountStore.getOrganisation()?.id}
                             />
                           )}
+                          {user && this.state.showGithubMessage && (
+                            <Row className={'px-3'}>
```
Comment (kyle-ssg) [inline frontend/web/components/App.js:500]: RE above comment, we can just put all the changes and markup into ``<GitHubMessage/>``

**Comment by kyle-ssg** [comment]
Reading through the purpose of this PR, I just want to make sure this is the best solution. There must be a way to handle this without UI?

E.g.

- Do GitHub Apps have WebHooks that detect uninstall ? If so we could invalidate the installation ID
- Is there a way to get a GH installation ID ? If so we could attempt to fetch it and invalidate if it doesnt exist

How does e.g. Vercel handle this ? 

Am I right in saying we could just subscribe to this https://docs.github.com/en/webhooks/webhook-events-and-payloads#github_app_authorization

File: frontend/web/components/modals/InvalidGithubIntegration.tsx (hunk 4 above)
Comment (novakzaballa) [reply frontend/web/components/modals/InvalidGithubIntegration.tsx:48]: That's for the case they want to keep the existing links between PR/Issue and feature flags, while they updated for some reason the GH App Installation ID, i.e. they removed the integration from GH accidentally. Maybe too much for an edge case?

**Comment by novakzaballa** [comment]
> Reading through the purpose of this PR, I just want to make sure this is the best solution. There must be a way to handle this without UI?
> 
> E.g.
> 
> * Do GitHub Apps have WebHooks that detect uninstall ? If so we could invalidate the installation ID
> * Is there a way to get a GH installation ID ? If so we could attempt to fetch it and invalidate if it doesnt exist
> 
> How does e.g. Vercel handle this ?
> 
> Am I right in saying we could just subscribe to this https://docs.github.com/en/webhooks/webhook-events-and-payloads#github_app_authorization

Yes, that is the plan but it was not part of the scope of this first version. However, maybe it is worth it to implement the webhook instead of this. Please let's discuss.

**Comment by novakzaballa** [comment]
> * Do GitHub Apps have WebHooks that detect uninstall ? If so we could invalidate the installation ID

Yep, 

> * Is there a way to get a GH installation ID ? If so we could attempt to fetch it and invalidate if it doesnt exist
> 
> How does e.g. Vercel handle this ?
> 
> Am I right in saying we could just subscribe to this https://docs.github.com/en/webhooks/webhook-events-and-payloads#github_app_authorization

>

**Pull request closed without merging**